### PR TITLE
EP Merge: Added abort check for expired special issue codes

### LINKED ISSUE #3048: Handle deactivated special issues in EP Merge app
Refinement notes:
- If there are deactivated special issues -- we might need an "abort" code in addition to success/fail.
- Not a need, but would make our metrics more clear. When we report results these get lumped into the "error" category, but deactivated special issues shouldn't be eligible for merging in the first place.
- Bigish lift ~1 week of work (integration + unit tests)
- We may not have a BIP endpoint to even check deactivated special issues (maybe a field in the contentions endpoint)
- In the past 2 weeks we've only had 6 jobs fail because of deactivated special issues.

### PR DESCRIPTION
**Pull request by dfitchett**
## What was the problem?

Currently, a merge job will fail if any of the EP400 contentions for merge have a special issue code that has a deactive date that is in the past, meaning the code is expired and will result in a 400 on the create contentions call to downstream BIP to add those contentions to the pending EP.

Associated tickets or Slack threads:
- #3048 

## How does this fix it?[^1]

This adds a check in the RUNNING_GET_EP400_CONTENTIONS state that verifies that the mergeable contentions do not have an expired deactive date.

## How to test this PR
- pull code
- Unit Test:
  - from /domain-ee/ee-ep-merge-app run `pytest`
- Integration Tests:
  - run [base platform](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Docker-Compose#platform-base)
  - from /domain-ee/ee-ep-merge-app run `pytest ./integration`
- End to End tests:
  - run this [CI](https://github.com/department-of-veterans-affairs/abd-vro/actions/workflows/ee-ep-merge-end-to-end.yml) action to verify successful integration test
  - run end to end tests locally:
    - `export BIP_CLAIM_URL=mock-bip-claims-api:20300`
    - run bgs and bip services and mocks:
      - `COMPOSE_PROFILES=“all” ./gradlew dockerComposeUp`
      - `COMPOSE_PROFILES="bip,bgs" ./gradlew -p mocks :dockerComposeUp`
    - `./gradlew :domain-ee:ee-ep-merge-app:endToEndTest`

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

### CODE REVIEW

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 1]
```diff
@@ -274,17 +279,49 @@ def on_get_ep400_contentions(self, event, pending_contentions_response=None):
...
+            request=request, hoppy_client=HOPPY.get_client(ClientName.GET_SPECIAL_ISSUE_TYPES), response_type=get_special_issue_types.Response
+        )
+
+        if response and response.status_code == 200:
+            deactive_special_issue_codes = [
+                special_issue.code
+                for special_issue in response.code_name_pairs or []
+                if special_issue.deactive_date and special_issue.deactive_date < datetime.now(special_issue.deactive_date.tzinfo)
+            ]
+            if deactive_special_issue_codes:
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]: Oof, this method gets really deeply nested. What if you negate the `if`s on lines 288 and here (294) to fail-fast? That will at least save on some indenting.

Still not sure I love the length though. Maybe the more nested code here can come out into a new method?

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 2]
```diff
@@ -274,17 +279,49 @@ def on_get_ep400_contentions(self, event, pending_contentions_response=None):
...
+                if deactive_codes_on_mergeable_contentions:
+                    self.add_job_error(
+                        f'EP400 claim #{self.job.ep400_claim_id} has contentions to merge that have the following deactive special issue codes: '
+                        f'{", ".join(deactive_codes_on_mergeable_contentions)}'
+                    )
+            else:
+                self.add_job_error(
+                    f'Could not retrieve special issue codes to verify EP400 claim #{self.job.ep400_claim_id} does not contain contentions with '
+                    f'deactive special issue codes'
+                )
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]: Do you not need to do a `self.send()` in this `@state.enter` method?

**Review by nanotone — COMMENTED** [review]
Let me know if you come up with a way to make that new method a bit more readable. If not, no worries.

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py (hunk 2 above)
Comment (dfitchett) [reply domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]: The method `verify_ep400_contention_special_issue_codes` is called inside of the `@running_get_ep400_contentions.enter` method which also calls `self.send()` at the bottom 👍

**Review by nanotone — APPROVED** [review]
Looks great now!

**Pull request merged**